### PR TITLE
Documentation UX revamp and internal CMake restructuring

### LINKED ISSUE #1905: Sort out and document CMake targets
We currently have a bunch of different `make` targets, which are partly the result of a) using CMake defaults, b) carrying over defaults from the time when the build system still used autotools, and c) custom targets (some of which have bad names for historic reasons).

### `all`
* compile all source files and link static and dynamic libraries, as well as the executables
* cythonize PyNEST and compile it;

### `install`
* run the [SLI help extractor](https://github.com/nest/nest-simulator/tree/master/extras/help_generator)

install everything to directories under the `<nest_install_dir>`
* [SLI standard library](https://github.com/nest/nest-simulator/tree/master/lib/sli), [SLI examples](https://github.com/nest/nest-simulator/tree/master/examples/nest), extracted SLI help files
* dynamic libraries, executables (`nest`, `sli`) and helpers (`nest-config`, `nest-vars.sh`)
* PyNEST and [its examples](https://github.com/nest/nest-simulator/tree/master/pynest/examples)
* the [testsuite](https://github.com/nest/nest-simulator/tree/master/testsuite)

### `install-nodoc`
* same as `install`, but without running the SLI help extractor

### `installcheck`
run the testsuite (essentially by calling [`do_tests.sh`](https://github.com/nest/nest-simulator/blob/master/testsuite/do_tests.sh)) using the installed executables and the test files from `<nest_install_dir>/share/doc/nest`

### `doc` and `fulldoc`
run Doxygen to extract the C++-level user documentation in either
* a [minimal configuration](https://github.com/nest/nest-simulator/blob/master/doc/normaldoc.conf.in) (`doc`), or in
* a [extended configuration](https://github.com/nest/nest-simulator/blob/master/doc/fulldoc.conf.in) (`fulldoc`) including class inheritance diagrams and all bling.

### `html`
run the Sphinx-based user-level documentation generation pipeline. This also runs the doc extractor that gets the model documentation snippets from the C++ files that implement them.

## Proposal

There is some (well justified) confusion between what these targets actually do. A proposal for a more sane naming and task split would probably be

* **`all`** can stay
* **`install`** would do the same as before minus running the doc-extractor
* **`userdoc`** would do what `html` currently does, plus run the SLI doc extractor
* **`devdoc`** would do what `fulldoc` currently does
* **`install-userdoc`** would install what **`userdoc`** has produced
* **`install-devdoc`** would install what **`devdoc`** has produced

This issue is also related to #291.

### PR DESCRIPTION
**Pull request by Helveg**
This PR aims to present users with a simple streamlined approach to building the documentation. A user requests which documentation they are interested in by toggling CMake flags for documentation types aimed at certain target audiences, and CMake figures out which documentation targets need to be configured for that. Documentation is generated either implicitly upon installation (`make install`), or explicitly by using `make docs`, or any of the subtargets, but the subtargets aren't considered public features.

Documentation targets are:

* `sphinxdocs`: Uses Sphinx to generate the html documentation, and lots of stuff happens, see comment in code.
* `doxygendocs`: Uses Doxygen to generate the dev docs.
* There is a legacy block for SLI docs, that I didn't refactor as SLI is removed soon.

Documentation flags are:

* `with-userdoc`: Toggles the user documentation, on by default. 
* `with-devdoc`: Toggles the developer documentation, off by default.

## User stories

* I want to install NEST
  * Userdocs are generated and installed along with the NEST library
* I want to turn off this documentation
  * I pass `-Dwith-userdoc=OFF`
* I want to have developer documentation
  * I pass `-Dwith-devdoc=ON`
* I want to build the docs, not NEST
  * I use `make docs` instead of `make install`

closes #1905

### CODE REVIEW

**Comment by Helveg** [comment]
Since building the user documentation brings in a lot of additional build time and dependencies, that need to be installed from different source (apt, pip, built from source, ...) and the ReadTheDocs documentation is more accessible, and more convenient, and we opted to turn off the local build of the documentation.

**Review by heplesser — APPROVED** [review]

**Comment by jessica-mitchell** [comment]
@Helveg thanks for this PR
the docs seem to build correctly but there is this error at the end of the for userdocs build:

``` 
cmake -Dwith-userdoc=ON ~/Work/repo/nest-simulator/
make docs
```

image



the devdoc flag does not elicit any errors.

**Comment by Helveg** [comment]
I can't reproduce this. Do you have the proper file permissions to execute this file, is it executable? If I'm interpreting the image correctly, it looks like it's trying to run `resolve_includes.py` as a shell script, which is not something that happens for me?

**Review by jessica-mitchell — CHANGES_REQUESTED** [review]
thanks @Helveg I just have a few changes

File: doc/htmldoc/developer_space/workflows/documentation_workflow/user_documentation_workflow.rst [hunk 1]
```diff
@@ -71,6 +71,12 @@ documentation and avoid using Conda, you can use pip:
 
     pip3 install -r <nest_source_dir>/doc/requirements.txt
 
+If you use pip, install ``pandoc`` from your platform's package manager (e.g. apt):
```
[the remark this answers is not in the document]
Comment (Helveg) [reply doc/htmldoc/developer_space/workflows/documentation_workflow/user_documentation_workflow.rst:74]: The `pandoc` that is available on `pip` does not contain any binaries, it appears to be a Python library only, it doesn't appear to have much in common with the CLI tool `pandoc` that we need here, other than the name and purpose.

**Review by jessica-mitchell — APPROVED** [review]
thanks lgtm!

File: doc/htmldoc/developer_space/workflows/documentation_workflow/user_documentation_workflow.rst (hunk 1 above)
[the remark this answers is not in the document]
Comment (jessica-mitchell) [reply doc/htmldoc/developer_space/workflows/documentation_workflow/user_documentation_workflow.rst:74]: ok, thanks

**Pull request merged**